sensitivity analysis/sensitivity_analysis.R

```diff
@@ -0,0 +1,193 @@
+params_fixed <- list()
+for(p in names(parameters)){
+  params_fixed[[p]] <- rep_len(parameters[[p]], length.out = length(runs))
+}
+
+# Combine the two lists
+runlist_final <- c(runlist, params_fixed)
+
+# Run the model
+output <- list()
+for(r in 1:length(runs)){
+  params <- lapply(runlist_final[-1], `[`, r)
+
+  output[[r]] <- as.data.frame(Run_Model(inputdata,
+                           derivs_V2_MM,
+                           params,
+                           num.years=NUM_YEARS,
+                           state=c(POM = 1, LMWC = 1, AGG = 1, MIC = 1, MAOM=1, CO2=0)))
+
+  # Assign run number column
+  output[[r]]$run_number <- r
+
+}
+
+# Combine data
+model_output <- bind_rows(output) %>% as_tibble()
+
+# Plot just the model outputs as a ribbon plot
+# Clean up data
+ribbon_plot_data <- model_output %>%
+  pivot_longer(cols = c(POM, LMWC, AGG, MIC, MAOM, CO2),
+               names_to = "metric",
+               values_to = "values") %>%
+  group_by(time, metric) %>%
+  summarize(sd = sd(values),
+            min = min(values),
+            max = max(values),
+            median = median(values))
+
+# Plot
+ribbon_plot <- ribbon_plot_data %>%
+  ggplot(aes(x = time, fill = metric)) +
+  geom_ribbon(aes(ymin = min, ymax = max),  alpha = 0.5) +
+  geom_line(aes(y = median, color = metric), linewidth = 1) +
+  facet_wrap(~metric, scales = "free_y") +
+  scale_color_brewer(palette = "BrBG") +
+  scale_fill_brewer(palette = "BrBG") +
+  labs(x = "Time (days)",
+       y = "Units?") +
+  theme_bw() ; ribbon_plot
+
+ggsave(plot = ribbon_plot, "./figures/model_outputs.jpg", height = 6, width = 9, units = "in")
+
+# Calculate relative importance
+# Define function
+calc_relimp <- function(x) {
+
+    t <- x$time # record time...
+    x$time <- NULL # and drop time column
+    run <- x$run_number[1] # record run number
+    x$run_number <- NULL # drop run number column
+    n <- x$name # record parameter name
+    x$name <- NULL # and drop
+
+    # Fit model. The formula notation means "CO2 as a
+    # function of all other terms" (which is why we dropped year)
+    lin <- lm(param ~ ., data = x)
+
+    # Calculate relative importance metrics and extract 'lmg' results
+    # 'lmg' is the R^2 contribution averaged over orderings among regressors;
+    # should sum to one because we're using relative importances (rela=TRUE)
+    # try()
+    relimp <- try(calc.relimp(lin, type = "lmg", rela = TRUE)@lmg)
+    if(class(relimp) == "try-error") {
+      message(t, " calc.relimp error")
+      return(NULL)
+    }
+
+    # Return a data frame: year, parameter, relative importance
+    tibble(time = unique(t),
+           output = names(relimp),
+           value = relimp,
+           source = n[1])
+}
+
+# Create dataframe to pass to relative importance function
+importance_test <- model_output %>%
+  left_join(bind_rows(runlist), by = "run_number")
+
+# Loop through each row of the data (except for the first day, where
```

Comment:
@leeyap I want to make sure I understand this comment correctly - so the first day is 0s and 1s because that is what we set as the starting state in line 64?

Reply:
Exactly! This is the default for Millenial but can be modified to run a steady state. The tutorial has a good example of this.

Reply:
Awesome!